src/backend/Program.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "Macros.h"
+#include "base/CCRef.h"
+#include "platform/CCPlatformMacros.h"
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+CC_BACKEND_BEGIN
+
+class ShaderModule;
+class Texture;
+
+class Program : public Ref
+{
+public:
+    struct TextureInfo
+    {
+        TextureInfo(const std::string& _name, const std::vector<uint32_t>& _indices, const std::vector<Texture*> _textures);
+        TextureInfo() = default;
+        ~TextureInfo();
+        TextureInfo& operator =(TextureInfo&& rhs);
+        
+        void retainTextures();
+        void releaseTextures();
+        
+        std::string name;
+        std::vector<uint32_t> indices;
+        std::vector<Texture*> textures;
+    };
+    
+    virtual int getUniformLocation(const std::string& uniform) = 0;
```

Comment:
please add const

Reply:
and i think should use getVertexUniformLoacation and getFragmentUniformLoacation instead